// promod/built/swd/driver/local.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.runLocalEnv = void 0;
const fs = require("fs");
const computeFsPaths = require("selenium-standalone/lib/compute-fs-paths");
const standAloneDefaultConfig = require("selenium-standalone/lib/default-config");
const getPort = require("get-port");
const remote_1 = require("selenium-webdriver/remote");
async function getOptsData(opts = {}) {
    const defaultConfig = standAloneDefaultConfig();
    if (!opts.version) {
        opts.version = defaultConfig.version;
    }
    if (opts.drivers) {
        opts.drivers = Object.keys(opts.drivers).reduce((config, driverName) => {
            config[driverName] = Object.assign({}, defaultConfig.drivers[driverName], opts.drivers[driverName]);
            return config;
        }, {});
    }
    else {
        opts.drivers = defaultConfig.drivers;
    }
    const fsPaths = await computeFsPaths({
        seleniumVersion: opts.version,
        drivers: opts.drivers,
        basePath: opts.basePath,
    });
    return fsPaths;
}
const throwInstructionError = (additional = '') => {
    throw new Error(`${additional}Run 'selenium-standalone install' to download browser drivers and selenium standalone server.`);
};
async function checkIfUpdateRequired(driverName, opts) {
    const optsData = await getOptsData(opts);
    if (driverName === 'selenium') {
        if (!fs.existsSync(optsData.selenium.installPath)) {
            throwInstructionError();
        }
    }
    if (driverName === 'chrome') {
        if (!fs.existsSync(optsData.chrome.installPath)) {
            throwInstructionError();
        }
    }
    if (driverName === 'firefox') {
        if (!fs.existsSync(optsData.firefox.installPath)) {
            throwInstructionError();
        }
    }
}
function checkIfDriverOrServerExists(pathTo) {
    if (!fs.existsSync(pathTo)) {
        throwInstructionError(`${pathTo} does not exists \n`);
    }
}
function mapToObject(map) {
    return Array.from(map).reduce((acc, [key, value]) => {
        acc[key] = value;
        return acc;
    }, {});
}
async function getCombinedConfig(config = {}) {
    const combinedConfig = config;
    if (!config.seleniumServerStartTimeout) {
        config.seleniumServerStartTimeout = 30000;
    }
    combinedConfig.capabilities =
        combinedConfig.capabilities.map_ instanceof Map
            ? mapToObject(combinedConfig.capabilities.map_)
            : combinedConfig.capabilities;
    if (!combinedConfig.seleniumServerJar) {
        const optsData = await getOptsData(config.selenium);
        checkIfUpdateRequired('selenium', config.selenium);
        const pathToServer = optsData.selenium.installPath;
        checkIfDriverOrServerExists(pathToServer);
        combinedConfig.seleniumServerJar = pathToServer;
    }
    if (!combinedConfig.chromeDriver && combinedConfig.capabilities.browserName === 'chrome') {
        const optsData = await getOptsData(config.selenium);
        checkIfUpdateRequired('chrome', config.selenium);
        const pathToChromedriver = optsData.chrome.installPath;
        checkIfDriverOrServerExists(pathToChromedriver);
        combinedConfig.chromeDriver = pathToChromedriver;
    }
    if (!combinedConfig.geckoDriver && combinedConfig.capabilities.browserName === 'firefox') {
        const optsData = await getOptsData(config.selenium);
        checkIfUpdateRequired('firefox', config.selenium);
        const pathTogeckoDriver = optsData.firefox.installPath;
        checkIfDriverOrServerExists(pathTogeckoDriver);
        combinedConfig.geckoDriver = pathTogeckoDriver;
    }
    return combinedConfig;
}
async function runLocalEnv(config) {
    if (config.seleniumAddress) {
        return config;
    }
    const combinedConfig = await getCombinedConfig(config);
    const serverConf = combinedConfig.localSeleniumStandaloneOpts || {};
    const port = combinedConfig.seleniumPort || (await getPort());
    if (!serverConf.args) {
        serverConf.args = combinedConfig.seleniumArgs || [];
    }
    if (!serverConf.jvmArgs) {
        serverConf.jvmArgs = combinedConfig.jvmArgs || [];
    }
    else if (!Array.isArray(serverConf.jvmArgs)) {
        throw new TypeError('jvmArgs should be an array.');
    }
    if (!serverConf.port) {
        serverConf.port = port;
    }
    if (combinedConfig.chromeDriver) {
        serverConf.jvmArgs.push(`-Dwebdriver.chrome.driver=${combinedConfig.chromeDriver}`);
    }
    if (combinedConfig.geckoDriver) {
        serverConf.jvmArgs.push(`-Dwebdriver.gecko.driver=${combinedConfig.geckoDriver}`);
    }
    const server = new remote_1.SeleniumServer(combinedConfig.seleniumServerJar, serverConf);
    await server.start(combinedConfig.seleniumServerStartTimeout);
    const address = await server.address();
    return {
        seleniumAddress: address,
        capabilities: combinedConfig.capabilities,
    };
}
exports.runLocalEnv = runLocalEnv;
